Use yesno arrays

benadida opened the pull request:
@eventualbuddha I think we need to include the possibility of a `['yes','no']` array, but otherwise I think this looks good. WDYT?

beausmith committed:
fix: Use standard locale string.

beausmith committed:
WIP

eventualbuddha commented on src/v1/index.ts:
```diff
@@ -88,7 +88,7 @@ function encodeBallotVotesInto(
       if (contest.type === 'yesno') {
         // yesno votes get a single bit
-        bits.writeBoolean(contestVote === 'yes')
+        bits.writeBoolean(contestVote === ['yes'])
```
You can't compare arrays like this, as this will always be `false`.

benadida committed:
updates for proper array comparisons

benadida committed:
added tests for error conditions

eventualbuddha commented on src/v0/index.ts:
```diff
@@ -101,18 +101,19 @@ function encodeYesNoVote(
-        )}`
-      )
+  if (
+    !(contestVote instanceof Array) ||
```
```suggestion
    !Array.isArray(contestVote) ||
```

eventualbuddha commented on src/v0/index.ts:
```diff
@@ -101,18 +101,19 @@ function encodeYesNoVote(
+    )
   }
+
+  return contestVote[0] === 'yes' ? '1' : '0'
```
This still implicitly encodes `[]` and `['no']` equivalently. Is that intended?

benadida replied:
oof, good point. How should we be encoding `[]`?

eventualbuddha replied:
It means "no vote", right? So I'm guessing for v0 it's just an empty string and for v1 it's an absent vote. I'll noodle on it.

eventualbuddha commented on src/v1/index.ts:
```diff
@@ -87,8 +88,22 @@ function encodeBallotVotesInto(
       if (contest.type === 'yesno') {
+        const ynVote = contestVote as YesNoVote
+
+        if (!(ynVote instanceof Array)) {
```
```suggestion
        if (!Array.isArray(ynVote)) {
```

eventualbuddha commented on src/v1/index.ts:
```diff
@@ -87,8 +88,22 @@ function encodeBallotVotesInto(
+
         // yesno votes get a single bit
-        bits.writeBoolean(contestVote === 'yes')
+        bits.writeBoolean(ynVote[0] === 'yes')
```
Same question as earlier: what about `[]`?

eventualbuddha reviewed (approved):
A question, and some minor fixes.

benadida committed:
better array checking

Co-authored-by: Brian Donovan <1938+eventualbuddha@users.noreply.github.com>

eventualbuddha committed:
fix: ensure empty votes are counted as such

When there is a record for a vote, but it's an empty array indicating "no votes", it should be treated the same as if it were not present.
